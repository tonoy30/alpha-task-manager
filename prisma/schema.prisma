datasource db {
  provider = "postgresql"
  url      = env("DB_URL")
}

generator client {
  provider = "prisma-client-js"
}

model User {
  id       String @id @default(uuid())
  username String
  password String
}
