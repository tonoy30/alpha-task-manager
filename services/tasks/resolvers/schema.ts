import { gql } from 'apollo-server'

export const typeDefs = gql`
  type Task {
    id: ID!
    title: String!
    status: Int!
    order: Int!
  }

  type List {
    id: ID!
    title: String!
    tasks: [Task]!
  }

  input CreateListInput {
    title: String!
  }

  input CreateTaskInput {
    title: String!
  }

  input UpdateTaskInput {
    title: String!
    status: Int!
  }

  input MoveTaskInput {
    previousPosition: Int!
    currentPosition: Int!
  }

  type MutationResult {
    success: Boolean!
  }

  type Query {
    lists: [List!]!
  }

  type Mutation {
    createList(input: CreateListInput!): List!
    createTask(listId: ID!, input: CreateTaskInput!): Task!
    updateTask(id: ID!, input: UpdateTaskInput!): Task
    moveTask(id: ID!, input: MoveTaskInput!): MutationResult!
  }
`
