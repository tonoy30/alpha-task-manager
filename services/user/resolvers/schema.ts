import { gql } from 'apollo-server'

export const typeDefs = gql`
  type User {
    id: ID!
    username: String!
  }

  input CreateUserInput {
    username: String!
    password: String!
  }

  input UpdateUserInput {
    username: String
  }

  type MutationResult {
    success: Boolean!
  }

  type Query {
    users: [User!]!
    user(id: ID!): User
  }

  type Mutation {
    createUser(input: CreateUserInput!): User!
    updateUser(id: ID!, input: UpdateUserInput!): User
    deleteUser(id: ID!): MutationResult!
  }
`
