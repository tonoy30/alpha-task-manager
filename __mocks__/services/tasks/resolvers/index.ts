import { List } from '@libs/model'
import { CreateListInput } from 'generated/types'

export const resolvers = {
  Query: {
    lists: () => queryData,
    pagedLists: (_parent: unknown, args: { skip: number; take?: number }) => {
      const { skip, take } = args
      const list = queryData.slice(skip, skip + (take ?? 2))
      return list
    },
  },
  Mutation: {
    createList: (
      _parent: unknown,
      { title }: CreateListInput,
      _ctx: unknown
    ) => {
      const newList = {
        id: '1',
        title,
        tasks: [],
      } as List
      mutationData.push(newList)
      return newList
    },
    createTask: () => {},
    updateTask: () => {},
    moveTask: () => {},
  },
}

export const queryData: List[] = [
  {
    id: '9cb80d31-49ed-4e6b-9eea-78f0b8a04eb8',
    title: 'List 1',
    tasks: [
      {
        id: 'bb97eb77-26d2-4d50-bfb9-a50003aabf25',
        position: 4,
        title: 'Task 4',
        status: 'Uncompleted',
      },
      {
        id: '6dda2e6f-4b58-4031-bff1-8cfe810dc77e',
        position: 3,
        title: 'Task 3',
        status: 'Uncompleted',
      },
      {
        id: 'd26423b6-80a1-455b-98e9-8f4bd068f2b7',
        position: 2,
        title: 'Task 2',
        status: 'Uncompleted',
      },
      {
        id: '8d995d0b-5de3-406e-8b7d-2e50f5f27274',
        position: 1,
        title: 'Task 1',
        status: 'Uncompleted',
      },
      {
        id: '528f55d5-3cb0-468b-a503-be21493c986b',
        position: 0,
        title: 'Task 0',
        status: 'Uncompleted',
      },
    ],
  },
  {
    id: '9cb80d31-49ed-4e6b-9eea-78f0b8a04eb9',
    title: 'List 2',
    tasks: [
      {
        id: 'bb97eb77-26d2-4d50-bfb9-a50003aabf25',
        position: 4,
        title: 'Task 4',
        status: 'Uncompleted',
      },
      {
        id: '6dda2e6f-4b58-4031-bff1-8cfe810dc77e',
        position: 3,
        title: 'Task 3',
        status: 'Uncompleted',
      },
      {
        id: 'd26423b6-80a1-455b-98e9-8f4bd068f2b7',
        position: 2,
        title: 'Task 2',
        status: 'Uncompleted',
      },
      {
        id: '8d995d0b-5de3-406e-8b7d-2e50f5f27274',
        position: 1,
        title: 'Task 1',
        status: 'Uncompleted',
      },
      {
        id: '528f55d5-3cb0-468b-a503-be21493c986b',
        position: 0,
        title: 'Task 0',
        status: 'Uncompleted',
      },
    ],
  },
  {
    id: '9cb80d31-49ed-4e6b-9eea-78f0b8a04e10',
    title: 'List 3',
    tasks: [
      {
        id: 'bb97eb77-26d2-4d50-bfb9-a50003aabf25',
        position: 4,
        title: 'Task 4',
        status: 'Uncompleted',
      },
      {
        id: '6dda2e6f-4b58-4031-bff1-8cfe810dc77e',
        position: 3,
        title: 'Task 3',
        status: 'Uncompleted',
      },
      {
        id: 'd26423b6-80a1-455b-98e9-8f4bd068f2b7',
        position: 2,
        title: 'Task 2',
        status: 'Uncompleted',
      },
      {
        id: '8d995d0b-5de3-406e-8b7d-2e50f5f27274',
        position: 1,
        title: 'Task 1',
        status: 'Uncompleted',
      },
      {
        id: '528f55d5-3cb0-468b-a503-be21493c986b',
        position: 0,
        title: 'Task 0',
        status: 'Uncompleted',
      },
    ],
  },
  {
    id: '9cb80d31-49ed-4e6b-9eea-78f0b8a04e11',
    title: 'List 4',
    tasks: [
      {
        id: 'bb97eb77-26d2-4d50-bfb9-a50003aabf25',
        position: 4,
        title: 'Task 4',
        status: 'Uncompleted',
      },
      {
        id: '6dda2e6f-4b58-4031-bff1-8cfe810dc77e',
        position: 3,
        title: 'Task 3',
        status: 'Uncompleted',
      },
      {
        id: 'd26423b6-80a1-455b-98e9-8f4bd068f2b7',
        position: 2,
        title: 'Task 2',
        status: 'Uncompleted',
      },
      {
        id: '8d995d0b-5de3-406e-8b7d-2e50f5f27274',
        position: 1,
        title: 'Task 1',
        status: 'Uncompleted',
      },
      {
        id: '528f55d5-3cb0-468b-a503-be21493c986b',
        position: 0,
        title: 'Task 0',
        status: 'Uncompleted',
      },
    ],
  },
]
export const mutationData: List[] = []
