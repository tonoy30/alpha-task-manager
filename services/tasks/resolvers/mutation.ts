import { Resolvers } from 'generated/types'
import { Context } from '../../../libs/context'

export const mutation: Resolvers<Context>['Mutation'] = {
  createUser: async (_parent, { input }, ctx) =>
    ctx.prisma.user.create({ data: input }),
  updateUser: async (_parent, { id, input }, ctx) =>
    ctx.prisma.user.update({
      where: { id },
      data: {
        username: input.username ?? undefined,
      },
    }),
}
