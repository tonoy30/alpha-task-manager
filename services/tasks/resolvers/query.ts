import { Resolvers } from 'generated/types'
import { Context } from '../../../libs/context'

export const query: Resolvers<Context>['Query'] = {
  users: async (_parent, _args, ctx) => ctx.prisma.user.findMany(),
  user: async (_parent, { id }, ctx) =>
    ctx.prisma.user.findUnique({
      where: { id },
    }),
}
