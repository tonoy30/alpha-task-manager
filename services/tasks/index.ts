import { ApolloServer } from 'apollo-server'
import { createGqlServer } from '../../libs/server'
import { resolvers } from './resolvers'
import { typeDefs } from './resolvers/schema'

export async function startServer(): Promise<void> {
  const server = await createGqlServer({
    typeDefs,
    resolvers,
  })

  const { url } = await server.listen(Number(process.env.USER_SERVICE_PORT))

  console.log(`User service running at ${url}`)
}
