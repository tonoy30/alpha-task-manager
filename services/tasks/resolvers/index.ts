import { Resolvers } from 'generated/types'
import { Context } from '../../../libs/context'
import { mutation } from './mutation'
import { query } from './query'

export const resolvers: Resolvers<Context> = {
  Query: query,
  Mutation: mutation,
}
