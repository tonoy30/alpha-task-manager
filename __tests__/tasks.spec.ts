import { addMocksToSchema } from '@graphql-tools/mock'
import { makeExecutableSchema } from '@graphql-tools/schema'
import { ApolloServer, gql } from 'apollo-server'
import { typeDefs } from '../services/tasks/resolvers/schema'
import { resolvers } from '../__mocks__/services/tasks/resolvers'

let testSever: ApolloServer

const listQuery = gql`
  query List {
    lists {
      id
      title
      tasks {
        id
        position
        title
        status
      }
    }
  }
`
const pagedListQuery = gql`
  query PagedListQuery($skip: Int!, $take: Int) {
    pagedLists(skip: $skip, take: $take) {
      id
      title
      tasks {
        id
        position
        title
        status
      }
    }
  }
`

beforeAll(() => {
  testSever = new ApolloServer({
    schema: addMocksToSchema({
      schema: makeExecutableSchema({ typeDefs, resolvers }),
      preserveResolvers: true,
    }),
  })
})

describe('Testing Task and List Queries', () => {
  it('fetches all the lists', async () => {
    const response = await testSever.executeOperation({ query: listQuery })
    expect(response).toBeTruthy()
    expect(response).toHaveProperty('data')
    expect(response.errors).toBeFalsy()
    expect(response).toMatchSnapshot()
  })
  it('fetches first 2 list', async () => {
    const response = await testSever.executeOperation({
      query: pagedListQuery,
      variables: {
        skip: 0,
        take: 2,
      },
    })
    expect(response).toBeTruthy()
    expect(response).toHaveProperty('data')
    expect(response.errors).toBeFalsy()
    expect(response).toMatchSnapshot()
  })
})
