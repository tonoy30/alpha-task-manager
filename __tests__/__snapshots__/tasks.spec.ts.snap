// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Task and List Queries fetches all the lists 1`] = `
{
  "data": {
    "lists": [
      {
        "id": "9cb80d31-49ed-4e6b-9eea-78f0b8a04eb8",
        "tasks": [
          {
            "id": "bb97eb77-26d2-4d50-bfb9-a50003aabf25",
            "position": 4,
            "status": "Uncompleted",
            "title": "Task 4",
          },
          {
            "id": "6dda2e6f-4b58-4031-bff1-8cfe810dc77e",
            "position": 3,
            "status": "Uncompleted",
            "title": "Task 3",
          },
          {
            "id": "d26423b6-80a1-455b-98e9-8f4bd068f2b7",
            "position": 2,
            "status": "Uncompleted",
            "title": "Task 2",
          },
          {
            "id": "8d995d0b-5de3-406e-8b7d-2e50f5f27274",
            "position": 1,
            "status": "Uncompleted",
            "title": "Task 1",
          },
          {
            "id": "528f55d5-3cb0-468b-a503-be21493c986b",
            "position": 0,
            "status": "Uncompleted",
            "title": "Task 0",
          },
        ],
        "title": "List 1",
      },
      {
        "id": "9cb80d31-49ed-4e6b-9eea-78f0b8a04eb9",
        "tasks": [
          {
            "id": "bb97eb77-26d2-4d50-bfb9-a50003aabf25",
            "position": 4,
            "status": "Uncompleted",
            "title": "Task 4",
          },
          {
            "id": "6dda2e6f-4b58-4031-bff1-8cfe810dc77e",
            "position": 3,
            "status": "Uncompleted",
            "title": "Task 3",
          },
          {
            "id": "d26423b6-80a1-455b-98e9-8f4bd068f2b7",
            "position": 2,
            "status": "Uncompleted",
            "title": "Task 2",
          },
          {
            "id": "8d995d0b-5de3-406e-8b7d-2e50f5f27274",
            "position": 1,
            "status": "Uncompleted",
            "title": "Task 1",
          },
          {
            "id": "528f55d5-3cb0-468b-a503-be21493c986b",
            "position": 0,
            "status": "Uncompleted",
            "title": "Task 0",
          },
        ],
        "title": "List 2",
      },
      {
        "id": "9cb80d31-49ed-4e6b-9eea-78f0b8a04e10",
        "tasks": [
          {
            "id": "bb97eb77-26d2-4d50-bfb9-a50003aabf25",
            "position": 4,
            "status": "Uncompleted",
            "title": "Task 4",
          },
          {
            "id": "6dda2e6f-4b58-4031-bff1-8cfe810dc77e",
            "position": 3,
            "status": "Uncompleted",
            "title": "Task 3",
          },
          {
            "id": "d26423b6-80a1-455b-98e9-8f4bd068f2b7",
            "position": 2,
            "status": "Uncompleted",
            "title": "Task 2",
          },
          {
            "id": "8d995d0b-5de3-406e-8b7d-2e50f5f27274",
            "position": 1,
            "status": "Uncompleted",
            "title": "Task 1",
          },
          {
            "id": "528f55d5-3cb0-468b-a503-be21493c986b",
            "position": 0,
            "status": "Uncompleted",
            "title": "Task 0",
          },
        ],
        "title": "List 3",
      },
      {
        "id": "9cb80d31-49ed-4e6b-9eea-78f0b8a04e11",
        "tasks": [
          {
            "id": "bb97eb77-26d2-4d50-bfb9-a50003aabf25",
            "position": 4,
            "status": "Uncompleted",
            "title": "Task 4",
          },
          {
            "id": "6dda2e6f-4b58-4031-bff1-8cfe810dc77e",
            "position": 3,
            "status": "Uncompleted",
            "title": "Task 3",
          },
          {
            "id": "d26423b6-80a1-455b-98e9-8f4bd068f2b7",
            "position": 2,
            "status": "Uncompleted",
            "title": "Task 2",
          },
          {
            "id": "8d995d0b-5de3-406e-8b7d-2e50f5f27274",
            "position": 1,
            "status": "Uncompleted",
            "title": "Task 1",
          },
          {
            "id": "528f55d5-3cb0-468b-a503-be21493c986b",
            "position": 0,
            "status": "Uncompleted",
            "title": "Task 0",
          },
        ],
        "title": "List 4",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": {
    "headers": Headers {
      Symbol(map): {},
    },
  },
}
`;

exports[`Testing Task and List Queries fetches first 2 list 1`] = `
{
  "data": {
    "pagedLists": [
      {
        "id": "9cb80d31-49ed-4e6b-9eea-78f0b8a04eb8",
        "tasks": [
          {
            "id": "bb97eb77-26d2-4d50-bfb9-a50003aabf25",
            "position": 4,
            "status": "Uncompleted",
            "title": "Task 4",
          },
          {
            "id": "6dda2e6f-4b58-4031-bff1-8cfe810dc77e",
            "position": 3,
            "status": "Uncompleted",
            "title": "Task 3",
          },
          {
            "id": "d26423b6-80a1-455b-98e9-8f4bd068f2b7",
            "position": 2,
            "status": "Uncompleted",
            "title": "Task 2",
          },
          {
            "id": "8d995d0b-5de3-406e-8b7d-2e50f5f27274",
            "position": 1,
            "status": "Uncompleted",
            "title": "Task 1",
          },
          {
            "id": "528f55d5-3cb0-468b-a503-be21493c986b",
            "position": 0,
            "status": "Uncompleted",
            "title": "Task 0",
          },
        ],
        "title": "List 1",
      },
      {
        "id": "9cb80d31-49ed-4e6b-9eea-78f0b8a04eb9",
        "tasks": [
          {
            "id": "bb97eb77-26d2-4d50-bfb9-a50003aabf25",
            "position": 4,
            "status": "Uncompleted",
            "title": "Task 4",
          },
          {
            "id": "6dda2e6f-4b58-4031-bff1-8cfe810dc77e",
            "position": 3,
            "status": "Uncompleted",
            "title": "Task 3",
          },
          {
            "id": "d26423b6-80a1-455b-98e9-8f4bd068f2b7",
            "position": 2,
            "status": "Uncompleted",
            "title": "Task 2",
          },
          {
            "id": "8d995d0b-5de3-406e-8b7d-2e50f5f27274",
            "position": 1,
            "status": "Uncompleted",
            "title": "Task 1",
          },
          {
            "id": "528f55d5-3cb0-468b-a503-be21493c986b",
            "position": 0,
            "status": "Uncompleted",
            "title": "Task 0",
          },
        ],
        "title": "List 2",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": {
    "headers": Headers {
      Symbol(map): {},
    },
  },
}
`;
